shiny/render/_data_frame_utils/_selection.py

```diff
@@ -177,274 +162,258 @@ class BrowserCellSelectionRegion(TypedDict):
+                )
 
 
 RowSelectionModeDeprecated = Literal["single", "multiple", "none", "deprecated"]
 
 
-def as_selection_mode(
-    selection_mode: SelectionMode,
+def as_selection_modes(
+    selection_modes: SelectionMode | Sequence[SelectionMode],
     *,
     name: str,
     editable: bool,
     row_selection_mode: RowSelectionModeDeprecated = "deprecated",
-) -> SelectionMode:
-    if row_selection_mode == "deprecated":
-        # Disable selection_mode if `editable=True``
-        if editable:
-            warnings.warn(
-                '`editable` can not be `True` while `selection_mode` != `"none"`. '
-                'Setting `selection_mode="none"`',
-                stacklevel=2,
-            )
-            return "none"
-
-        return selection_mode
-
-    warn_deprecated(
-        f"`{name}(row_selection_mode=)` has been superseded by `{name}(selection_mode=)`."
-        f' Please use `{name}(mode="{row_selection_mode}_row")` instead.'
-    )
+) -> SelectionModes:
+    # If a user supplied `row_selection_mode`, use the value but warn against it
+    if row_selection_mode != "deprecated":
+
+        warn_deprecated(
+            f"`{name}(row_selection_mode=)` has been superseded by `{name}(selection_modes=)`."
+            f' Please use `{name}(mode="{row_selection_mode}_row")` instead.'
+        )
+
+        if row_selection_mode == "none":
+            selection_modes = ("none",)
+        elif row_selection_mode == "single":
+            selection_modes = ("row",)
+        elif row_selection_mode == "multiple":
+            selection_modes = ("rows",)
+        else:
+            raise ValueError("Unknown row_selection_mode: {row_selection_mode}")
 
-    if row_selection_mode == "none":
-        return as_selection_mode("none", name=name, editable=editable)
-    elif row_selection_mode == "single":
-        return as_selection_mode("row", name=name, editable=editable)
-    elif row_selection_mode == "multiple":
-        return as_selection_mode("rows", name=name, editable=editable)
+    # Upgrade to tuple
+    if isinstance(selection_modes, str):
+        selection_modes_tuple = (selection_modes,)
     else:
-        raise ValueError("Unknown row_selection_mode: {row_selection_mode}")
+        assert isinstance(selection_modes, Sequence)
+        selection_modes_tuple = tuple(selection_modes)
+
+    # Disable selection_mode if `editable=True``
+    if editable and "none" not in selection_modes_tuple:
+        warnings.warn(
+            '`editable` can not currently be `True` while `selection_mode` != `"none"`. '
+            'Setting `selection_mode=("none",)`',
+            stacklevel=3,
+        )
+        selection_modes_tuple = ("none",)
+
+    # Trim values
+    if "none" in selection_modes_tuple:
+        return ("none",)
+    if "row" in selection_modes_tuple and "rows" in selection_modes_tuple:
+        selection_modes_tuple = tuple(
+            mode for mode in selection_modes_tuple if mode != "row"
+        )
```

Comment:
I think this will result in some slightly odd behavior :

- If the input is `(,)`, then this returns `(,)`
- If the input is `("rows", "none")`, then this returns `("none",)`
- If the input is `("rows", "row", "rows")`, then this returns `("rows", "rows")`

Also a small nit: it's inconsistent that the none code path has an immediate return, but the row+rows code path sets a value that's returned later.

Something like this will be more robust, and it will disallow unknown values. It also will mean that the explanation of resolution rules can be simpler:

```py
selection_modes_set = set(selection_modes)

if not selection_modes_set.issubset({"none", "row", "rows", "col", "cols"}):
    raise ValueError( ... )

if len(selection_modes_set) == 0:
    return ("none",)

if "none" in selection_modes_set and len(selection_modes_set) > 1:
    raise ValueError( ... )

if "row" in selection_modes_set and "rows" in selection_modes_set:
    raise ValueError( ... )

if "col" in selection_modes_set and "cols" in selection_modes_set:
    raise ValueError( ... )

return tuple(s)
```

Although, maybe the internal representation for no selection _should_ just be an empty tuple?

Reply:
One more thought: I think that the "correct" data structure for selection modes actually looks like this:

```py
@dataclass
class SelectionModes:
    row: Literal["none", "single", "multiple"]
    col: Literal["none", "single", "multiple"]
    rectangle: Literal["none", "multiple"]  # This one might need better names
```

Then there's no wiggle room in the type system (no empty tuples, no duplicate entries, etc) and it makes some of the code slightly cleaner -- instead of searching in a tuple for `"rows"`, you can index directly to it, as in `s.row == "multiple"`.

That said, since this is small and internal-facing, the exact representation isn't super important, and we can always change it in the future. But to me this feels like the right way to represent this information.

Reply:
Using new class structure in python and TS

Currently using `rect: Literal["cell", "region"]` within the class